File: src/ml_abnormal_detection.R
# One-hot encording in R
# ============================================================================
# Environment setting ----
# ============================================================================
if(!requireNamespace("pacman", quietly = T)) install.packages("pacman")
if(!requireNamespace("reticulate", quietly = T)) install.packages("reticulate")
reticulate::use_condaenv("DAJIN")
pacman::p_load(tidyverse, keras, reticulate, caret, e1071, gridExtra, svglite, progress)
# reticulate::py_config()

np <- import("numpy")
pd <- import("pandas")
model_selection <- import("sklearn.model_selection")
tf <- import("tensorflow")
np_utils <- import("keras.utils.np_utils")
regularizers <- import("keras.regularizers")
ftools <- import("functools")
# <<<<<<<<<<

args <- commandArgs(trailingOnly = TRUE)
file_name <- args[1]
# file_name <- ".DAJIN_temp/data/DAJIN.txt"

output_dir <- ".DAJIN_temp/results"
output_npz <- file_name %>% str_replace(".txt*$", ".npz")
output_model <- file_name %>% str_replace(".txt*$", ".h5")
output_file_name <- file_name %>% str_remove(".*/") %>% str_remove(".txt.*$")

df <- read_tsv(file_name, col_names = c("ID", "SEQ", "LABEL"), col_types = cols())

df$SEQ <- str_c("MIDS=", df$SEQ, sep = "")
df_sim <- df %>% filter(str_detect(LABEL,"simulate"))
df_real <- df %>% filter(!str_detect(LABEL,"simulate"))

# ============================================================================
# # one-hot encoding ----
# ============================================================================
one_hot_encode <- function(seq){
  seq_len <- seq[1] %>% str_count()
  array_oh <- array(0, dim = c(length(seq), seq_len, 4))
  #
  list_seq <- seq %>%
    str_split("")
  #
  pb <- progress_bar$new(
      format = "  One-hot encording... [:bar] :percent eta: :eta",
      total = length(list_seq), clear = FALSE, width = 100)
  #
  for(i in seq_along(list_seq)){
    pb$tick()
    x <- list_seq[[i]]
    for(j in seq_along(x)){
      if(x[j] == "M") array_oh[i,j,] <- c(1,0,0,0)
      else if(x[j] == "I") array_oh[i,j,] <- c(0,1,0,0)
      else if(x[j] == "D") array_oh[i,j,] <- c(0,0,1,0)
      else if(x[j] == "S") array_oh[i,j,] <- c(0,0,0,1)
    }
  }
  return(array_oh)
}
# # ////
X_all <- one_hot_encode(df$SEQ)
#
X_sim <- X_all[str_detect(df$LABEL, "simulate"), , ]
X_real <- X_all[!str_detect(df$LABEL, "simulate"), , ]
#
# # ////
# np_load = ftools$partial(np$load, allow_pickle=TRUE)
# npz = np_load(output_npz)

# X_all = npz["X_all"]
# X_sim = npz["X_sim"]
# X_real = npz["X_real"]

labels_categorical <- pd$factorize(df_sim$LABEL)[[1]] %>%
  np_utils$to_categorical()
labels_id <- pd$factorize(df_sim$LABEL)[[2]] %>% str_remove("_aligned_reads")

train_test <- model_selection$train_test_split(
  X_sim, labels_categorical,
  test_size = 0.2, shuffle = TRUE
)

X_train <- train_test[[1]]
X_test <- train_test[[2]]
Y_train <- train_test[[3]]
Y_test <- train_test[[4]]

# ============================================================================
# model construction ---- 
# ============================================================================
# os <- import("os")
# os$environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

model <- keras_model_sequential()
model %>%
  layer_conv_1d(
    filters = 32, kernel_size = 32, activation = "relu",
    input_shape = c(dim(X_train)[2], 4),
    name = "1st_Conv1D"
  ) %>%
  layer_max_pooling_1d(pool_size = 4, name = "1st_MaxPooling1D") %>%
  layer_conv_1d(
    filters = 32, kernel_size = 16,
    activation = "relu", name = "2nd_Conv1D"
  ) %>%
  layer_max_pooling_1d(pool_size = 4, name = "2nd_MaxPooling1D") %>%
  layer_conv_1d(
    filters = 32, kernel_size = 8,
    activation = "relu", name = "3rd_Conv1D"
  ) %>%
  layer_max_pooling_1d(pool_size = 4, name = "3rd_MaxPooling1D") %>%
  layer_conv_1d(
    filters = 32, kernel_size = 4,
    activation = "relu", name = "4th_Conv1D"
  ) %>%
  layer_max_pooling_1d(pool_size = 4, name = "4th_MaxPooling1D") %>%
  layer_flatten(name = "flatten") %>%
  layer_dense(64, activation = "relu", name = "1st_FC")

# L2-norm
alpha <- c(0.1) # hyperparameter
model %>% layer_dense(64,
  activation = "linear",
  activity_regularizer = regularizers$l2(alpha), name = "L2-softmax"
)
#
model %>% layer_dense(length(labels_id), activation = "softmax", name = "final_layer")
# summary(model)

model %>% compile(
  optimizer = "adam",
  loss = "categorical_crossentropy",
  metrics = c("accuracy")
)
# <<<<<<<<<<<<<

print("Model fitting...")
suppressMessages(
    history <- model %>% fit(X_train, Y_train,
    epochs = 20, verbose = 0,
    validation_split = 0.2, shuffle = TRUE
  )
)
# plot(history)

# ============================================================================
# Abnormal allele detection ---------------
# ============================================================================

cosine_sim <- function(a, b) crossprod(a, b) / sqrt(crossprod(a) * crossprod(b))

get_cosine_score <- function(model, train, test){
  model_ <- keras_model(model$get_layer(index = 0L)$input,
                        model$get_layer(index = -2L)$output)
  # print(summary(model_))
  train_vector <- model_ %>% predict(train, verbose = 0)
  test_vector <- model_ %>% predict(test, verbose = 0)
  cosine_sim(train_vector[1,], test_vector[1,])
  # train_list <- split(train_vector, rep(1:nrow(train_vector), each = ncol(train_vector)))
  # score <- parallel::mclapply(train_list, function(x) {
  #   apply(x, 1, function(y) cosine_sim(test_vector[i, ], y)) %>% max()
  # })
  score <- rep(0, nrow(test_vector))
  pb <- progress_bar$new(
    format = "  Cosine similarity... [:bar] :percent eta: :eta",
    total = length(score), clear = FALSE, width = 100
  )
  for(i in seq_along(score)){
    pb$tick()
    score[i] <- apply(train_vector, 1, function(x) cosine_sim(test_vector[i, ], x)) %>% max()
    }
  return(score)
}
# cos_max <- function(train_vector, test_vector){
#   score <- apply(train_vector, 1, function(x) cosine_sim(test_vector[i, ], x)) %>% max()
#   return(score)
#   }

cos_all <- get_cosine_score(model, X_train[0:500, , ], X_all)
#
df_all <- tibble(barcode = df$LABEL, cos_similarity = cos_all)
df_all <- df_all %>% mutate(label = if_else(str_detect(barcode, "simulate"), "simulated", "real")) # -----------------------------
cos_threshold <- df_all %>%
  filter(label == "simulated") %>%
  summarize(quantile(cos_similarity, 0.001)) %>%
  unlist()

# ============================================================================
# Output the results ---------------
# ============================================================================

# Save one-hot encording ...
np$savez_compressed(output_npz,
  X_all = X_all,
  X_sim = X_sim,
  X_real = X_real
)
# Save the model ...
model$save(output_model)
# Save Anomaly annotation ...
df_anomaly <- df_all %>%
  filter(label == "real") %>%
  mutate(abnormal_prediction = if_else(cos_similarity > cos_threshold, "normal", "abnormal")) %>%
  bind_cols(., df_real) %>%
  select(ID, barcode, abnormal_prediction)
#
write_tsv(df_anomaly, ".DAJIN_temp/data/DAJIN_anomaly_classification.txt", col_names = F)
# Save labels
write_tsv(data.frame(labels_id), ".DAJIN_temp/data/DAJIN_anomaly_classification_labels.txt", col_names = F)

# confusion matrix for classification -----------------------
predicted_labels <- model %>% predict_classes(X_test)
df_id <- data.frame(label = seq_along(labels_id) - 1, id = labels_id)
df_pred <- data.frame(label = predicted_labels)
predicted_labels <- inner_join(df_id, df_pred, by = "label") %>%
  .$id %>%
  as.factor()

df_true <- apply(Y_test, 1, which.max) %>% data.frame(label = . - 1)
true_labels <- inner_join(df_id, df_true, by = "label") %>%
  .$id %>%
  as.factor()

cm <- confusionMatrix(predicted_labels, true_labels)
cm <- data.frame(cm$table)

plotTable <- cm %>%
  mutate(goodbad = ifelse(cm$Prediction == cm$Reference, "correct", "incorrect")) %>%
  group_by(Reference) %>%
  mutate(prop = Freq / sum(Freq))
# plotTable

g_cm <- ggplot(plotTable, aes(x = Reference, y = Prediction, fill = goodbad, alpha = prop)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%.2f", prop)), vjust = .5, fontface = "bold", alpha = 1) +
  scale_fill_manual(values = c(correct = "#FF4B00", incorrect = "#5290FF")) +
  xlim(rev(levels(cm$Reference))) +
  theme_bw(base_size = 15, base_family = "Arial") +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.text.x = element_text(angle = 30, hjust = 1)
  )
ggsave(sprintf("%s/confusion_matrix.jpg", output_dir), g_cm, width = 8, height = 6)

# Cosine similarity...
g_cos <- ggplot(df_all, aes(x = barcode, y = cos_similarity)) +
  geom_boxplot(aes(color = label)) +
  geom_hline(yintercept = cos_threshold, linetype = "dashed") +
  expand_limits(y = 0) +
  labs(x = "Sample ID", y = "Cosine similarity") +
  coord_flip() +
  scale_x_discrete(limits = rev(levels(as.factor(df_all$barcode)))) +
  theme_bw(base_size = 15) +
  theme(
    plot.title = element_text(hjust = 0.5),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )
ggsave(sprintf("%s/cosine_similarity.jpg", output_dir), g_cos, width = 10, height = 10)

File: src/ml_abnormal_detection.py
import os
os.environ['TF_CPP_MIN_LOG_LEVEL']='3'
os.environ['TF_FORCE_GPU_ALLOW_GROWTH']='true'

import sys
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, LabelBinarizer
from sklearn.neighbors import LocalOutlierFactor

import tensorflow as tf
from tensorflow.keras.layers import Conv1D, Dense, Flatten,MaxPooling1D

################################################################################
#! I/O naming
################################################################################

#===========================================================
#? TEST auguments
#===========================================================

file_name = ".DAJIN_temp/data/DAJIN_MIDS.txt"
ont_cont = "barcode01"
mutation_type = "I"
threads = 10

#===========================================================
#? Auguments
#===========================================================

args = sys.argv
file_name = args[1]
ont_cont = args[2]
mutation_type = args[3]
threads = int(args[4])

if ont_cont == "" :
    raise ValueError("control is empty")

if mutation_type == "" :
    raise ValueError("mutation_type is empty")

if threads == "" :
    threads = 1

#===========================================================
#? Input
#===========================================================

df = pd.read_csv(file_name, header=None, sep='\t')
df.columns = ["seqID", "seq", "barcodeID"]
df.seq = "MIDS=" + df.seq

df_cont = df[df.barcodeID.str.contains(ont_cont)].reset_index(drop=True)
df_cont.barcodeID = "wt_simulated"

df_sim = df[df.barcodeID.str.contains("simulated")].reset_index(drop=True)
df_sim = df_sim.append(df_cont).reset_index(drop=True)
df_real = df[~df.barcodeID.str.contains("simulated")].reset_index(drop=True)

del df
del df_cont

################################################################################
#! Novelity (Anomaly) detection
################################################################################

#===========================================================
#? Encording
#===========================================================

def label_encorde_seq(seq):
    label_seq = seq.apply(list).\
        apply(LabelEncoder().fit_transform).\
        apply(pd.Series).\
        to_numpy()
    return(label_seq)

#===========================================================
#? Train test split
#===========================================================

labels, labels_index = pd.factorize(df_sim.barcodeID)

X_train, X_test, Y_train, Y_test = train_test_split(
    label_encorde_seq(df_sim.seq), labels,
    test_size=0.2, shuffle=True)

#===========================================================
#? LOF
#===========================================================
print("LOF") #? >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
clf = LocalOutlierFactor(n_neighbors=20, leaf_size = 400, novelty=True, n_jobs = threads)

clf.fit(X_train[:10000])

outliers = clf.predict(label_encorde_seq(df_real.seq))
outliers = np.where(outliers==1, "normal", "abnormal") 
df_real["outliers"] = outliers

# df_real.groupby("barcodeID").outliers.value_counts()
df_real.to_csv(
    ".DAJIN_temp/data/DAJIN_MIDS_prediction_result.txt",
    header=False, index=False, sep="\t")

if mutation_type == "P" :
    sys.exit()
    
################################################################################
#! Classification
################################################################################

#===========================================================
#? Encording
#===========================================================

def onehot_encode_seq(seq):
    onehot_seq = np.apply_along_axis(LabelBinarizer().fit_transform, 1, label_encorde_seq(seq))
    return(onehot_seq)

#===========================================================
#? Train test split
#===========================================================

labels = tf.keras.utils.to_categorical(labels)

X_train, X_test, Y_train, Y_test = train_test_split(
    onehot_encode_seq(df_sim.seq), labels,
    test_size=0.2, shuffle=True)

#===========================================================
#? CNN
#===========================================================

model = tf.keras.Sequential()

model.add(Conv1D(filters=32, kernel_size=32, activation="relu", padding = "same",
                input_shape=(X_train.shape[1], X_train.shape[2]), name="1st_Conv1D"))
model.add(MaxPooling1D(pool_size=4, name="1st_MaxPooling1D"))

model.add(Conv1D(filters=64, kernel_size=16, padding = "same",
                activation="relu", name="2nd_Conv1D"))
model.add(MaxPooling1D(pool_size=8, name="2nd_MaxPooling1D"))

model.add(Conv1D(filters=128, kernel_size=8, padding = "same",
                activation="relu", name="3rd_Conv1D"))
model.add(MaxPooling1D(pool_size=16, name="3rd_MaxPooling1D"))

model.add(Flatten(name="flatten"))

model.add(Dense(64, activation='relu', name="1st_FC"))

model.add(Dense(len(labels_index), activation='softmax', name="final_layer"))
model.compile(loss=tf.keras.losses.categorical_crossentropy,
              optimizer=tf.keras.optimizers.Adam(),
              metrics=['accuracy'])

model.fit(X_train, Y_train,
          epochs=15, verbose=1, batch_size = 64,
          validation_data=(X_test, Y_test), shuffle=True)

#===========================================================
#? prediction
#===========================================================
from tqdm import tqdm

X_real = onehot_encode_seq(df_real.seq)
del df_real["seq"] #! >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

iter_ = 1000
prediction = np.zeros(X_real.shape[0], dtype="uint8")
for i in tqdm(range(0, X_real.shape[0], iter_)):
    predict_ = model.predict(X_real[i: i + iter_].astype("float16"),
                             verbose=0, batch_size=32)
    prediction[i:i+iter_] = np.argmax(predict_, axis=1)

del X_real #! >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
    
# prediction = model.predict(onehot_encode_seq(df_real.seq).astype(np.float16))
# prediction = prediction.argmax(axis=1)

df_real["prediction"] = prediction
df_real["prediction"].mask(df_real["outliers"] == "abnormal", "abnormal", inplace=True)
del df_real["outliers"] #! >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

for index,label in enumerate(labels_index):
    label=label.replace("_simulated","")
    df_real["prediction"].mask(df_real["prediction"] == index, label, inplace=True)

print(df_real.groupby("barcodeID").prediction.value_counts()) #? >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

df_real.to_csv(
    ".DAJIN_temp/data/DAJIN_MIDS_prediction_result.txt",
    header=False, index=False, sep="\t")




# # ====================================
# # RandomForestClassifier
# # ====================================
# from sklearn.ensemble import RandomForestClassifier
# clf = RandomForestClassifier(n_jobs = threads)
# clf.fit(X_train, Y_train)

# # test = df_real.seqID == "00222820-ad67-4774-b5f9-8fda0fba4c83"
# # test = df_real.seqID == "00ea11a6-9c44-4d42-84e7-e03a113c6cea"
# # test = X_real[test]
# # clf.predict(test)

# prediction = clf.predict(X_real)

# df_real["outliers"] = outliers
# df_real["prediction"] = prediction
# df_real["prediction"].mask(df_real["outliers"] == "abnormal", "abnormal", inplace=True)
# del df_real["outliers"] #!--------------------------------

# for index,label in enumerate(labels_index):
#     label=label.replace("_simulated","")
#     df_real["prediction"].mask(df_real["prediction"] == index, label, inplace=True)

# df_real.groupby("barcodeID").prediction.value_counts()

# df_real.to_csv(
#     ".DAJIN_temp/data/DAJIN_MIDS_prediction_result.txt",
#     header=False, index=False, sep="\t")


# =============================================================
# PCA
# =============================================================
# X_all = label_encorde_seq(df.seq)
# X_all = onehot_seq

# import numpy as np
# from sklearn.decomposition import PCA

# pca = PCA(n_components=5)
# X_all = pca.fit(X_all).transform(X_all)
# X_all = X_all * pca.explained_variance_ratio_

# X_sim = X_all[df.barcodeID.str.contains(ont_cont) + df.barcodeID.str.contains("simulated")]
# X_real = X_all[~df.barcodeID.str.contains("simulated")]

# nsamples, nx, ny = X_all.shape
# X_all_reshape = X_all.reshape((nsamples,nx*ny))
# X_sim = X_all_reshape[df.barcodeID.str.contains(ont_cont) | df.barcodeID.str.contains("simulated")]
# X_real = X_all_reshape[~df.barcodeID.str.contains("simulated")]


# =============================================================
# LGBMClassifier
# =============================================================
# import lightgbm as lgb
# from sklearn.metrics import accuracy_score

# model = lgb.LGBMClassifier(n_jobs=threads)
# model.fit(X_train, Y_train)

# prediction = model.predict_proba(X_test)
# prediction = np.argmax(prediction, axis=1)

# accuracy = accuracy_score(prediction, Y_test)
# print('Averaging accuracy:', accuracy)

# prediction = model.predict_proba(X_real)
# prediction = np.argmax(prediction, axis=1)

# df_real["prediction"] = prediction
# df_real["prediction"].mask(df_real["outliers"] == "abnormal", "abnormal", inplace=True)
# del df_real["outliers"]

# for index,label in enumerate(labels_index):
#     label=label.replace("_simulated","")
#     df_real["prediction"].mask(df_real["prediction"] == index, label, inplace=True)


# df_real.groupby("barcodeID").prediction.value_counts()

# df_real.to_csv(
#     ".DAJIN_temp/data/DAJIN_MIDS_prediction_result.txt",
#     header=False, index=False, sep="\t")

# ====================================
# IsolationForest
# ====================================
# from sklearn.ensemble import IsolationForest
# clf = IsolationForest(n_jobs=10) 

# train = X_train[0:10000]
# clf.fit(train)

# y_pred_outliers = clf.predict(X_real)

# df_real["outliers"] = y_pred_outliers

# df_real.groupby("barcodeID").outliers.value_counts()


# # Save labels
# pd.Series(labels_index).to_csv(
#     '.DAJIN_temp/data/DAJIN_anomaly_classification_labels.txt',
#     header=False, index=False, sep="\t")



# ====================================
# PCA for X_real
# ====================================
# import numpy as np
# from sklearn.decomposition import PCA

# pca = PCA(n_components=10)
# X_real = pca.fit(X_real).transform(X_real)
# print(pca.explained_variance_ratio_)

# colors = ['navy', 'turquoise', 'darkorange']
# lw = 2

# y, barcode = pd.factorize(df_real.barcodeID)

# import matplotlib.pyplot as plt
# plt.figure()
# for color, i, bar in zip(colors, [0, 1, 2], barcode):
#     plt.scatter(X_r[y == i, 0], X_r[y == i, 1], color=color, alpha=.8, lw=lw, label=bar)

# plt.legend(loc='best', shadow=False, scatterpoints=1)
# plt.savefig("test.png")

# ====================================
# Neural net
# ====================================

# from sklearn.neural_network import MLPClassifier
# clf = MLPClassifier(solver='lbfgs', hidden_layer_sizes=(5,2))
# clf.fit(X_train, Y_train)
# clf.score(X_test, Y_test)

# prediction = clf.predict_proba(X_real)
# prediction = np.argmax(prediction, axis=1)

# df_real["outliers"] = outliers
# df_real["prediction"] = prediction
# df_real["prediction"].mask(df_real["outliers"] == "abnormal", "abnormal", inplace=True)

# for index,label in enumerate(labels_index):
#     label=label.replace("_simulated","")
#     df_real["prediction"].mask(df_real["prediction"] == index, label, inplace=True)


# df_real.groupby("barcodeID").prediction.value_counts()
